File: web/src/application/wallGeometry.ts
// Plan-space wall footprints with mitered corner joins and T-joins.
//
// A wall's footprint is the rectangle around its axis. When exactly two
// wall ends meet at a point, both footprints are mitered: each side face
// is extended to its intersection with the matching side face of the
// other wall, so the pair shares the two miter corners and the seam runs
// between them. Collinear continuations, star joints (3+ ends) and
// too-sharp angles (miter limit) keep square butt ends.
//
// A wall end with no coincident end that lands on another wall's axis
// segment is a T-join: its end face is trimmed (or extended) to butt
// against the near face of the continuous wall, which stays unbroken.

import type { Point3D, WallDatum } from "../domain/project";

export type Point2 = [number, number];

const JOIN_TOLERANCE = 1e-4;
const PARALLEL_EPS = 1e-9;
const MITER_LIMIT_FACTOR = 4;

interface Line2 {
  point: Point2;
  dir: Point2;
}

interface WallEnd {
  wall: WallDatum;
  endpoint: "start" | "end";
}

function samePoint(a: Point3D, b: Point3D): boolean {
  return (
    Math.abs(a[0] - b[0]) <= JOIN_TOLERANCE &&
    Math.abs(a[1] - b[1]) <= JOIN_TOLERANCE &&
    Math.abs(a[2] - b[2]) <= JOIN_TOLERANCE
  );
}

function intersectLines(first: Line2, second: Line2): Point2 | null {
  const cross = first.dir[0] * second.dir[1] - first.dir[1] * second.dir[0];
  if (Math.abs(cross) < PARALLEL_EPS) {
    return null;
  }
  const dx = second.point[0] - first.point[0];
  const dy = second.point[1] - first.point[1];
  const t = (dx * second.dir[1] - dy * second.dir[0]) / cross;
  return [first.point[0] + t * first.dir[0], first.point[1] + t * first.dir[1]];
}

/** Direction from the given endpoint into the wall body, and its left normal. */
function outgoingFrame(wall: WallDatum, endpoint: "start" | "end"): { o: Point2; nl: Point2 } {
  const sign = endpoint === "start" ? 1 : -1;
  const dx = wall.end[0] - wall.start[0];
  const dy = wall.end[1] - wall.start[1];
  const length = Math.hypot(dx, dy);
  const o: Point2 = [(sign * dx) / length, (sign * dy) / length];
  return { o, nl: [-o[1], o[0]] };
}

function sideLines(wall: WallDatum, endpoint: "start" | "end"): { left: Line2; right: Line2 } {
  const joint = wall[endpoint];
  const { o, nl } = outgoingFrame(wall, endpoint);
  const half = wall.thickness / 2;
  return {
    left: { point: [joint[0] + nl[0] * half, joint[1] + nl[1] * half], dir: o },
    right: { point: [joint[0] - nl[0] * half, joint[1] - nl[1] * half], dir: o },
  };
}

/** Every other wall end coinciding with this one. */
function endPartners(
  wall: WallDatum,
  endpoint: "start" | "end",
  walls: readonly WallDatum[],
): WallEnd[] {
  const joint = wall[endpoint];
  const others: WallEnd[] = [];
  for (const candidate of walls) {
    if (candidate.id === wall.id) continue;
    for (const candidateEnd of ["start", "end"] as const) {
      if (samePoint(candidate[candidateEnd], joint)) {
        others.push({ wall: candidate, endpoint: candidateEnd });
      }
    }
  }
  return others;
}

/**
 * Miter corners against the single coincident wall end, in the outgoing
 * frame: [cornerLeft, cornerRight], or null when the end stays square.
 */
function miterCorners(
  wall: WallDatum,
  endpoint: "start" | "end",
  partner: WallEnd,
): [Point2, Point2] | null {
  const own = sideLines(wall, endpoint);
  const other = sideLines(partner.wall, partner.endpoint);
  const cornerLeft = intersectLines(own.left, other.right);
  const cornerRight = intersectLines(own.right, other.left);
  if (!cornerLeft || !cornerRight) return null;
  const joint = wall[endpoint];
  const limit = MITER_LIMIT_FACTOR * Math.max(wall.thickness, partner.wall.thickness);
  const reach = Math.max(
    Math.hypot(cornerLeft[0] - joint[0], cornerLeft[1] - joint[1]),
    Math.hypot(cornerRight[0] - joint[0], cornerRight[1] - joint[1]),
  );
  if (reach > limit) return null;
  return [cornerLeft, cornerRight];
}

/**
 * T-join corners: the end butts against the near face of a wall whose axis
 * segment passes under this endpoint. [cornerLeft, cornerRight] in the
 * outgoing frame, or null when no continuous wall qualifies.
 */
function tButtCorners(
  wall: WallDatum,
  endpoint: "start" | "end",
  walls: readonly WallDatum[],
): [Point2, Point2] | null {
  const joint = wall[endpoint];
  const { o } = outgoingFrame(wall, endpoint);

  let best: { face: Line2; distance: number; thickness: number } | null = null;
  for (const candidate of walls) {
    if (candidate.id === wall.id) continue;
    const ax = candidate.end[0] - candidate.start[0];
    const ay = candidate.end[1] - candidate.start[1];
    const axisLength = Math.hypot(ax, ay);
    if (axisLength === 0) continue;
    const dir: Point2 = [ax / axisLength, ay / axisLength];
    const normal: Point2 = [-dir[1], dir[0]];
    const half = candidate.thickness / 2;
    const relX = joint[0] - candidate.start[0];
    const relY = joint[1] - candidate.start[1];
    const along = relX * dir[0] + relY * dir[1];
    // Interior of the axis segment only; end vicinities belong to corner logic.
    if (along < half || along > axisLength - half) continue;
    const offset = relX * normal[0] + relY * normal[1];
    if (Math.abs(offset) > half + JOIN_TOLERANCE) continue;
    // The face on the side of the continuous wall that this wall's body faces.
    const side = o[0] * normal[0] + o[1] * normal[1];
    if (Math.abs(side) < 1e-6) continue; // parallel walls
    const sign = side > 0 ? 1 : -1;
    const face: Line2 = {
      point: [
        candidate.start[0] + normal[0] * sign * half,
        candidate.start[1] + normal[1] * sign * half,
      ],
      dir,
    };
    const distance = Math.abs(offset);
    if (!best || distance < best.distance) {
      best = { face, distance, thickness: candidate.thickness };
    }
  }
  if (!best) return null;

  const own = sideLines(wall, endpoint);
  const cornerLeft = intersectLines(own.left, best.face);
  const cornerRight = intersectLines(own.right, best.face);
  if (!cornerLeft || !cornerRight) return null;
  const limit = MITER_LIMIT_FACTOR * Math.max(wall.thickness, best.thickness);
  const reach = Math.max(
    Math.hypot(cornerLeft[0] - joint[0], cornerLeft[1] - joint[1]),
    Math.hypot(cornerRight[0] - joint[0], cornerRight[1] - joint[1]),
  );
  if (reach > limit) return null;
  return [cornerLeft, cornerRight];
}

/** Join corners for one wall end: corner miter, T-butt, or null (square). */
function endJoinCorners(
  wall: WallDatum,
  endpoint: "start" | "end",
  walls: readonly WallDatum[],
): [Point2, Point2] | null {
  const partners = endPartners(wall, endpoint, walls);
  if (partners.length === 1) {
    return miterCorners(wall, endpoint, partners[0]);
  }
  if (partners.length > 1) {
    return null;
  }
  return tButtCorners(wall, endpoint, walls);
}

/**
 * Plan footprint of a wall as [startLeft, endLeft, endRight, startRight],
 * left/right relative to the start→end direction.
 */
export function wallFootprint(wall: WallDatum, walls: readonly WallDatum[]): Point2[] {
  const dx = wall.end[0] - wall.start[0];
  const dy = wall.end[1] - wall.start[1];
  const length = Math.hypot(dx, dy);
  const n: Point2 = [-dy / length, dx / length];
  const half = wall.thickness / 2;
  let startLeft: Point2 = [wall.start[0] + n[0] * half, wall.start[1] + n[1] * half];
  let startRight: Point2 = [wall.start[0] - n[0] * half, wall.start[1] - n[1] * half];
  let endLeft: Point2 = [wall.end[0] + n[0] * half, wall.end[1] + n[1] * half];
  let endRight: Point2 = [wall.end[0] - n[0] * half, wall.end[1] - n[1] * half];

  const startJoin = endJoinCorners(wall, "start", walls);
  if (startJoin) {
    // Outgoing dir at the start is start→end, so outgoing-left is footprint-left.
    [startLeft, startRight] = startJoin;
  }
  const endJoin = endJoinCorners(wall, "end", walls);
  if (endJoin) {
    // Outgoing dir at the end is reversed, so outgoing-left is footprint-right.
    [endRight, endLeft] = endJoin;
  }
  return [startLeft, endLeft, endRight, startRight];
}

File: webim/domain/wall_geometry.py
"""Plan-space wall footprints with joins and opening pieces.

Direct port of web/src/application/wallGeometry.ts so the Blender adapter
renders exactly the same solids as WeBIM Web: mitered corner joins, butt
joins (older wall runs through), T-joins trimmed to the crossed wall's
near face, and walls decomposed into pieces around their openings.

Pure Python — no bpy, no ifcopenshell.
"""

from __future__ import annotations

import math
from dataclasses import dataclass

from .project import NativeWall

Point2 = tuple[float, float]

JOIN_TOLERANCE = 1e-4
PARALLEL_EPS = 1e-9
MITER_LIMIT_FACTOR = 4


@dataclass(frozen=True, slots=True)
class WallPiece:
    """One solid fragment: a plan polygon spanning [z_bottom, z_top]."""

    corners: tuple[Point2, ...]
    z_bottom: float
    z_top: float


def _join_type_at(wall: NativeWall, endpoint: str) -> str:
    return wall.join_start if endpoint == "start" else wall.join_end


def _point_of(wall: NativeWall, endpoint: str):
    return wall.start if endpoint == "start" else wall.end


def _same_point(a, b) -> bool:
    return all(abs(x - y) <= JOIN_TOLERANCE for x, y in zip(a, b))


def _intersect(p1: Point2, d1: Point2, p2: Point2, d2: Point2) -> Point2 | None:
    cross = d1[0] * d2[1] - d1[1] * d2[0]
    if abs(cross) < PARALLEL_EPS:
        return None
    dx = p2[0] - p1[0]
    dy = p2[1] - p1[1]
    t = (dx * d2[1] - dy * d2[0]) / cross
    return (p1[0] + t * d1[0], p1[1] + t * d1[1])


def _axis_frame(wall: NativeWall):
    ax = wall.end[0] - wall.start[0]
    ay = wall.end[1] - wall.start[1]
    length = math.hypot(ax, ay)
    if length == 0:
        return None
    direction = (ax / length, ay / length)
    return direction, (-direction[1], direction[0]), length


def _outgoing(wall: NativeWall, endpoint: str):
    sign = 1.0 if endpoint == "start" else -1.0
    frame = _axis_frame(wall)
    direction = frame[0]
    o = (sign * direction[0], sign * direction[1])
    return o, (-o[1], o[0])


def _side_lines(wall: NativeWall, endpoint: str):
    joint = _point_of(wall, endpoint)
    o, nl = _outgoing(wall, endpoint)
    half = wall.thickness / 2
    left = ((joint[0] + nl[0] * half, joint[1] + nl[1] * half), o)
    right = ((joint[0] - nl[0] * half, joint[1] - nl[1] * half), o)
    return left, right


def _end_partners(wall: NativeWall, endpoint: str, walls):
    joint = _point_of(wall, endpoint)
    partners = []
    for candidate in walls:
        if candidate.id == wall.id:
            continue
        for candidate_end in ("start", "end"):
            if _same_point(_point_of(candidate, candidate_end), joint):
                partners.append((candidate, candidate_end))
    return partners


def _cut_end_at_face(wall, endpoint, face_point, face_dir, other_thickness):
    (left_p, left_d), (right_p, right_d) = _side_lines(wall, endpoint)
    corner_left = _intersect(left_p, left_d, face_point, face_dir)
    corner_right = _intersect(right_p, right_d, face_point, face_dir)
    if corner_left is None or corner_right is None:
        return None
    joint = _point_of(wall, endpoint)
    limit = MITER_LIMIT_FACTOR * max(wall.thickness, other_thickness)
    reach = max(
        math.hypot(corner_left[0] - joint[0], corner_left[1] - joint[1]),
        math.hypot(corner_right[0] - joint[0], corner_right[1] - joint[1]),
    )
    if reach > limit:
        return None
    return corner_left, corner_right


def _miter_corners(wall, endpoint, partner):
    partner_wall, partner_end = partner
    own_left, own_right = _side_lines(wall, endpoint)
    other_left, other_right = _side_lines(partner_wall, partner_end)
    corner_left = _intersect(own_left[0], own_left[1], other_right[0], other_right[1])
    corner_right = _intersect(own_right[0], own_right[1], other_left[0], other_left[1])
    if corner_left is None or corner_right is None:
        return None
    joint = _point_of(wall, endpoint)
    limit = MITER_LIMIT_FACTOR * max(wall.thickness, partner_wall.thickness)
    reach = max(
        math.hypot(corner_left[0] - joint[0], corner_left[1] - joint[1]),
        math.hypot(corner_right[0] - joint[0], corner_right[1] - joint[1]),
    )
    if reach > limit:
        return None
    return corner_left, corner_right


def _butt_corner_corners(wall, endpoint, partner, walls):
    partner_wall, _ = partner
    frame = _axis_frame(partner_wall)
    if frame is None:
        return None
    direction, normal, _ = frame
    o, _ = _outgoing(wall, endpoint)
    side = o[0] * normal[0] + o[1] * normal[1]
    if abs(side) < 1e-6:
        return None
    sign = 1.0 if side > 0 else -1.0
    ids = [candidate.id for candidate in walls]
    runs_through = ids.index(wall.id) < ids.index(partner_wall.id)
    face_sign = -sign if runs_through else sign
    half = partner_wall.thickness / 2
    face_point = (
        partner_wall.start[0] + normal[0] * face_sign * half,
        partner_wall.start[1] + normal[1] * face_sign * half,
    )
    return _cut_end_at_face(wall, endpoint, face_point, direction, partner_wall.thickness)


def _t_join_target(wall, endpoint, walls):
    joint = _point_of(wall, endpoint)
    o, _ = _outgoing(wall, endpoint)
    best = None
    best_distance = None
    for candidate in walls:
        if candidate.id == wall.id:
            continue
        if abs(candidate.start[2] - wall.start[2]) > JOIN_TOLERANCE:
            continue
        frame = _axis_frame(candidate)
        if frame is None:
            continue
        direction, normal, length = frame
        half = candidate.thickness / 2
        rel_x = joint[0] - candidate.start[0]
        rel_y = joint[1] - candidate.start[1]
        along = rel_x * direction[0] + rel_y * direction[1]
        if along < half or along > length - half:
            continue
        offset = rel_x * normal[0] + rel_y * normal[1]
        if abs(offset) > half + JOIN_TOLERANCE:
            continue
        side = o[0] * normal[0] + o[1] * normal[1]
        if abs(side) < 1e-6:
            continue
        if best is None or abs(offset) < best_distance:
            best = candidate
            best_distance = abs(offset)
    return best


def _t_butt_corners(wall, endpoint, walls):
    target = _t_join_target(wall, endpoint, walls)
    if target is None:
        return None
    direction, normal, _ = _axis_frame(target)
    o, _ = _outgoing(wall, endpoint)
    side = o[0] * normal[0] + o[1] * normal[1]
    sign = 1.0 if side > 0 else -1.0
    half = target.thickness / 2
    face_point = (
        target.start[0] + normal[0] * sign * half,
        target.start[1] + normal[1] * sign * half,
    )
    return _cut_end_at_face(wall, endpoint, face_point, direction, target.thickness)


def _end_join_corners(wall, endpoint, walls):
    own_type = _join_type_at(wall, endpoint)
    if own_type == "SQUARE":
        return None
    partners = _end_partners(wall, endpoint, walls)
    if len(partners) == 1:
        partner = partners[0]
        other_type = _join_type_at(partner[0], partner[1])
        if other_type == "SQUARE":
            return None
        if own_type == "BUTT" or other_type == "BUTT":
            return _butt_corner_corners(wall, endpoint, partner, walls)
        return _miter_corners(wall, endpoint, partner)
    if len(partners) > 1:
        return None
    return _t_butt_corners(wall, endpoint, walls)


def wall_footprint(wall: NativeWall, walls) -> tuple[Point2, ...]:
    """Plan polygon [start_left, end_left, end_right, start_right]."""
    frame = _axis_frame(wall)
    if frame is None:
        return ()
    _, n, _ = frame
    half = wall.thickness / 2
    start_left = (wall.start[0] + n[0] * half, wall.start[1] + n[1] * half)
    start_right = (wall.start[0] - n[0] * half, wall.start[1] - n[1] * half)
    end_left = (wall.end[0] + n[0] * half, wall.end[1] + n[1] * half)
    end_right = (wall.end[0] - n[0] * half, wall.end[1] - n[1] * half)

    start_join = _end_join_corners(wall, "start", walls)
    if start_join is not None:
        start_left, start_right = start_join
    end_join = _end_join_corners(wall, "end", walls)
    if end_join is not None:
        end_right, end_left = end_join
    return (start_left, end_left, end_right, start_right)


def wall_pieces(wall: NativeWall, walls) -> tuple[WallPiece, ...]:
    """Extrudable fragments around openings, mitered ends preserved."""
    footprint = wall_footprint(wall, walls)
    frame = _axis_frame(wall)
    if frame is None or not footprint:
        return ()
    direction, normal, length = frame
    half = wall.thickness / 2

    def cut_pair(t: float):
        return (
            (
                wall.start[0] + direction[0] * t + normal[0] * half,
                wall.start[1] + direction[1] * t + normal[1] * half,
            ),
            (
                wall.start[0] + direction[0] * t - normal[0] * half,
                wall.start[1] + direction[1] * t - normal[1] * half,
            ),
        )

    def pair_at(t: float):
        if t <= 0:
            return footprint[0], footprint[3]
        if t >= length:
            return footprint[1], footprint[2]
        return cut_pair(t)

    def polygon(a: float, b: float):
        a_left, a_right = pair_at(a)
        b_left, b_right = pair_at(b)
        return (a_left, b_left, b_right, a_right)

    spans = sorted(
        (
            (
                max(0.0, opening.offset - opening.width / 2),
                min(length, opening.offset + opening.width / 2),
                opening,
            )
            for opening in wall.openings
            if opening.offset + opening.width / 2
            > max(0.0, opening.offset - opening.width / 2)
        ),
        key=lambda item: item[0],
    )

    pieces: list[WallPiece] = []
    cursor = 0.0
    for start, end, opening in spans:
        if start > cursor:
            pieces.append(WallPiece(polygon(cursor, start), 0.0, wall.height))
        if opening.sill_height > 0:
            pieces.append(WallPiece(polygon(start, end), 0.0, opening.sill_height))
        head = opening.sill_height + opening.height
        if head < wall.height:
            pieces.append(WallPiece(polygon(start, end), head, wall.height))
        cursor = max(cursor, end)
    if cursor < length:
        pieces.append(WallPiece(polygon(cursor, length), 0.0, wall.height))
    return tuple(pieces)
